docs/external_objectives.rst

```diff
@@ -0,0 +1,318 @@
+example, which can be used as a template for wrapping other codes. Finally, an example
+script is shown for how to run the TERPSICHORE optimization with multiprocessing.
+
+Simple example
+--------------
+
+The following is a simple example of a custom function that is not JAX transformable.
+It was adapted from the test ``test_external_vs_generic_objectives`` in
+``tests/test_examples.py``.
+
+The function must take a single positional argument, which can be either a single
+Equilibrium or a list of Equilibria. Additional inputs can be passed as keyword
+arguments. In this example, the function returns the three scalar quatities
+:math:`\langle\beta\rangle_{vol}`, :math:`\langle\beta\rangle_{pol}`, and
+:math:`\langle\beta\rangle_{tor}`. It writes these quantities to a NetCDF file and then
+reads them back from the file. This mimics I/O from the common VMEC "wout" format, and
+also demonstrates an operation that is not automatically differentiable.
+
+::
+
+    def beta_from_file(eq, path=""):
+        # write data
+        file = Dataset(path, mode="w", format="NETCDF3_64BIT_OFFSET")
+        data = eq.compute(["<beta>_vol", "<beta_pol>_vol", "<beta_tor>_vol"])
+        betatotal = file.createVariable("betatotal", np.float64)
+        betatotal[:] = data["<beta>_vol"]
+        betapol = file.createVariable("betapol", np.float64)
+        betapol[:] = data["<beta_pol>_vol"]
+        betator = file.createVariable("betator", np.float64)
+        betator[:] = data["<beta_tor>_vol"]
+        file.close()
+        # read data
+        file = Dataset(path, mode="r")
+        betatot = float(file.variables["betatotal"][0])
+        betapol = float(file.variables["betapol"][0])
+        betator = float(file.variables["betator"][0])
+        file.close()
+        return np.atleast_1d([betatot, betapol, betator])
+
+This function can then be used in an optimization with the ``ExternalObjective`` class.
+This objective must be initialized with the Equilibrium to be optimized, ``eq``, and the
+following other arguments:
+
+* ``fun``, the external function. In this example it is the function ``beta_from_file``.
+* ``dim_f``, the dimension of the output of the function ``fun``. In this example
+  ``dim_f=3`` because ``beta_from_file`` returns an array of size 3.
+* ``vectorized``, whether ``fun`` expects a single Equilibrium or a list of
+  Equilibria. Since the function ``beta_from_file`` takes a single Equilibrium as its
+  only positional argument, ``vectorized=False``.
+
+All other input parameters that the external function requires can be passed as keyword
+arguments to ``ExternalObjective``. In this example, ``path`` specifies the file name
+of the NetCDF file that the function writes/reads. Since ``vectorized=False``, the
+function ``beta_from_file`` will be evaluated sequentially when computing the finite
+differences (rather than in parallel), so the same file will be overwritten each time
+for the different Equilibria.
```

Comment:
the different equilibria part is bc for the finite diffs, a different equilibrium is solved with slightly perturbed bdry and passed to here right? might want to make that explicit

Reply:
The `ExternalObjective`/`TERPSICHORE` objectives don't "re-solve" the equilibrium, they only evaluate the objective for the equilibrium parameters they are passed. Just like how `QuasisymmetryBoozer` or any other objectives don't solve the equilibrium themselves either. If you want to make sure it is a good equilibrium, you need to include the `ForceBalance` objective as a constraint like usual.